litani 1.20.0 (new formula)

ronakfof opened the pull request:
AWS Build Accumulator collects build jobs from multiple sources before
executing them concurrently. It provides platform-independent job
control (timeouts, return code control) and an output format that is
easy to render into reports. AWS Build Accumulator shines where your
project uses multiple different build systems or requires an unified
interface describing heterogeneous build jobs.
https://github.com/awslabs/aws-build-accumulator

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

ronakfof commented:
### Running `brew audit --strict litani` give the following output. What is the best way to get past this?
```
ronakfof/litani/litani:
  * Non-executables were installed to "/usr/local/opt/litani/bin".
    The offending files are:
      /usr/local/opt/litani/bin/test
      /usr/local/opt/litani/bin/bin
      /usr/local/opt/litani/bin/lib
      /usr/local/opt/litani/bin/templates
      /usr/local/opt/litani/bin/doc
```

ronakfof commented:
### On building using `brew install --build-from-source litani.rb` fails with the following even though `jinja` is installed using `resources` - 

Fails for both
```
system "cd #{bin}/doc && #{Formula["python@3.10"].opt_bin/"python3"} configure"
system "cd #{bin}/doc && ninja"
```

```
==> cd /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc && /home/linuxbrew/.linuxbrew/opt/python@3.10/bin/python3 configure
Last 15 lines from /home/ronakf/.cache/Homebrew/Logs/litani/06.python3:
2022-01-25 17:59:03 +0000

cd /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc && /home/linuxbrew/.linuxbrew/opt/python@3.10/bin/python3 configure

Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/configure", line 20, in <module>
    import jinja2
ModuleNotFoundError: No module named 'jinja2'
```
How to resolve this dependency issue?

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,59 @@
+class Litani < Formula
```
Is this supposed to be `AwsBuildAccumulator`?

ronakfof replied:
I would like to keep it as `Litani` as that is the name widely used for this package.

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,59 @@
+  depends_on "mandoc" => :build
+  depends_on "scdoc" => :build
+  depends_on "gnuplot"
+  depends_on "ninja"
```
This is needed at runtime?

ronakfof replied:
Both `mandoc` and `scdoc` are just required to run these 2 commands in the formula - 
```
    system "cd #{bin}/doc && #{Formula["python@3.9"].opt_bin/"python3"} configure"
    system "cd #{bin}/doc && ninja"
```

carlocab replied:
I was asking about `ninja`, since it is clear that `mandoc` and `scdoc` are build-only.

ronakfof replied:
Yes, `ninja` is needed at runtime.

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,59 @@
+      r.stage do
+        system Formula["python@3.10"].opt_bin/"python3", *Language::Python.setup_install_args(prefix)
+      end
+    end
```
We shouldn't install these into the `prefix`. Most Python formulae install their Python dependencies in a private directory; typically a virtualenv, to avoid install conflicts.

ronakfof replied:
I tried doing that initially but I was either getting import errors.

And In order to use `virtualenv_install_with_resources`, I need to add `setup.py` and need to change the directory structure which is not ideal.

ronakfof replied:
On installing using - 
```
class Litani < Formula
  include Language::Python::Virtualenv
...
...
  def install
    venv = virtualenv_create(libexec, Formula["python@3.9"].opt_bin/"python3")
    venv.pip_install resources
    bin.install "lib", "doc", "bin", "test", "templates", "litani"
    bin.install_symlink "#{libexec}/bin"
  end
```
Got this error after installation on running `litani -h`
```
(22-01-25 19:06:33) <0> [~]
Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/bin/litani", line 34, in <module>
    from lib import litani, litani_report, ninja_syntax
  File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/lib/litani_report.py", line 116, in <module>
    class MemoryTraceRenderer:
  File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/lib/litani_report.py", line 117, in MemoryTraceRenderer
    jinja_env: jinja2.Environment
AttributeError: module 'jinja2' has no attribute 'Environment'
```

ronakfof replied:
I tried - 
```
xy = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python#{xy}/site-packages"
resources.each do |r|
  r.stage do
    system Formula["python@3.9"].opt_bin/"python3", *Language::Python.setup_install_args(libexec/"vendor)
  end
end
```
And then running still `litani` failed with - 
```
Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/bin/litani", line 34, in <module>
    from lib import litani, litani_report, ninja_syntax
  File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/lib/litani_report.py", line 116, in <module>
    class MemoryTraceRenderer:
  File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/lib/litani_report.py", line 117, in MemoryTraceRenderer
    jinja_env: jinja2.Environment
AttributeError: module 'jinja2' has no attribute 'Environment'
```

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,59 @@
+        system Formula["python@3.10"].opt_bin/"python3", *Language::Python.setup_install_args(prefix)
+      end
+    end
+    bin.install "lib", "doc", "bin", "test", "templates", "litani"
```
Why are all these going into `bin`?

ronakfof replied:
The structure of a working directory is [here](https://github.com/awslabs/aws-build-accumulator).
`litani` file uses import statements like `import lib.litani_report.py` and more, which can only work if they are in same directory.

ronakfof replied:
> ### On building using `brew install --build-from-source litani.rb` fails with the following even though `jinja` is installed using `resources` -
> 
> Fails `system "cd #{bin}/doc && #{Formula["python@3.10"].opt_bin/"python3"} configure"` with -
> 
> ```
> ==> cd /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc && /home/linuxbrew/.linuxbrew/opt/python@3.10/bin/python3 configure
> Last 15 lines from /home/ronakf/.cache/Homebrew/Logs/litani/06.python3:
> 2022-01-25 17:59:03 +0000
> 
> cd /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc && /home/linuxbrew/.linuxbrew/opt/python@3.10/bin/python3 configure
> 
> Traceback (most recent call last):
>   File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/configure", line 20, in <module>
>     import jinja2
> ModuleNotFoundError: No module named 'jinja2'
> ```
> 
> And `system "cd #{bin}/doc && ninja"` with -
> 
> ```
> Last 15 lines from /home/ronakf/.cache/Homebrew/Logs/litani/07.doc:
> AttributeError: module 'yaml' has no attribute 'safe_load'
> [18/22] converting litani-outcome-table.json.roff HTML
> [19/22] converting litani-run.json.roff HTML
> [20/22] Giving unique header IDs to litani-outcome-table.json.html
> [21/22] Giving unique header IDs to litani-run.json.html
> [22/22] Building final HTML page
> FAILED: /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/out/html/index.html
> /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/bin/build-html-doc --html-manuals /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-init.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-run-build.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-dump-run.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-add-job.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-run.json.html /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/html_unique/litani-outcome-table.json.html --template-dir /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/templates  --out-file /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/out/html/index.html  --roff-html-dir /home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/tmp/roff_to_html
> Traceback (most recent call last):
>   File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/bin/build-html-doc", line 102, in <module>
>     main()
>   File "/home/linuxbrew/.linuxbrew/Cellar/litani/1.19.0/bin/doc/bin/build-html-doc", line 86, in main
>     env = jinja2.Environment(
> AttributeError: module 'jinja2' has no attribute 'Environment'
> ninja: build stopped: subcommand failed.
> ```
> 
> How to resolve this dependency issue?

NOTE: If the following code is removed, formula seems to be working fine - 
```
    cd bin/"doc" do
      system Formula["python@3.9"].opt_bin/"python3", "configure"
      system "ninja"
    end
    man.install Dir["#{bin}/doc/out/man/*.roff"]
    doc.install Dir["#{bin}/doc/out/html/index.html"]
```

carlocab replied:
Does man read `*.roff` files? That's still an unusual format for manpages.

ronakfof replied:
> Why are all these going into bin?

Resolved this issue

> Does man read *.roff files? That's still an unusual format for manpages.

Yes. Will update this. Thanks.

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,58 @@
+      end
+    end
+    bin.install "lib", "doc", "bin", "test", "templates", "litani"
+    system "cd #{bin}/doc && #{Formula["python@3.9"].opt_bin/"python3"} configure"
+    system "cd #{bin}/doc && ninja"
```
```suggestion
    cd bin/"doc" do
      system "python3", "configure"
      system "ninja"
    end
```
Though, these really shouldn't be in `bin`.

ronakfof replied:
What is the alternative since - 

> The structure of a working directory is here.
litani file uses import statements like import lib.litani_report.py and more, which can only work if they are in same directory.

ronakfof replied:
I installed the folders into `lib` and `share`. Then i created symlinks for all those using `bin.install_symlink ..`
Is that an ideal approach?

carlocab replied:
No, install everything in `libexec` and add symlinks or env scripts (see [`write_env_script`](https://rubydoc.brew.sh/Pathname.html#write_env_script-instance_method) and [`env_script_all_files`](https://rubydoc.brew.sh/Pathname.html#env_script_all_files-instance_method)) as appropriate.

You probably need to set `PYTHONPATH` for `litani` to find the privately installed Python modules.

ronakfof replied:
Resolved with
```
    libexec.install Dir["*"]
    bin.install_symlink libexec/"litani"
```

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,58 @@
+  end
+
+  test do
+    system "which", "litani"
+    system Formula["python@3.9"].opt_bin/"python3", "-c", "import jinja2"
+    system Formula["python@3.9"].opt_bin/"python3", "-c", "import voluptuous"
```
```suggestion
```

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,58 @@
+    system "which", "litani"
+    system Formula["python@3.9"].opt_bin/"python3", "-c", "import jinja2"
+    system Formula["python@3.9"].opt_bin/"python3", "-c", "import voluptuous"
+    system "#{bin}/test/run", "--fast", "--output-dir", buildpath
```
`buildpath` doesn't exist anymore. `bin` should not contain directories. We should also aim for a test the simulates user interaction rather than running unit tests.

ronakfof replied:
`#{bin}/test/run` contains `end-to-end` tests which verifies complete functionality on the underlying system.

SMillerDev replied:
We don't want to run full end to end tests though, we just want to run a single command that a user would run to see if that works. End to end tests are for upstream

ronakfof replied:
Resolved

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,72 @@
+    lib.install "lib", "bin"
+    share.install "doc", "templates"
+
+    bin.install_symlink lib/"lib", lib/"bin", share/"templates"
```
We shouldn't be installing symlinks into `bin` unless they're to executable files.

ronakfof replied:
> Resolved with
> 
>     libexec.install Dir["*"]
>     bin.install_symlink libexec/"litani"
>

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,72 @@
+    bin.install_symlink lib/"lib", lib/"bin", share/"templates"
+    share.install_symlink lib/"lib"
+
+    cd share/"doc" do
+      system Formula["python@3.9"].opt_bin/"python3", "configure"
+      system "ninja"
+    end
```
Can we do this in the build directory (your current working directory; you can also refer to this as `buildpath`) instead of in the formula's prefix?

ronakfof replied:
updated this to - 
```
    libexec.install Dir["*"]
    bin.install_symlink libexec/"litani"

    cd libexec/"doc" do
      system Formula["python@3.9"].opt_bin/"python3", "configure"
      system "ninja"
    end
    man.install Dir[libexec/"doc/out/man/*.roff"]
    doc.install Dir[libexec/"doc/out/html/index.html"]
```
And it seems to be working fine now.

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,69 @@
+  end
+
+  def install
+    xy = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
+    ENV.prepend_create_path "PYTHONPATH", lib/"python#{xy}/site-packages"
+    resources.each do |r|
+      r.stage do
+        system Formula["python@3.9"].opt_bin/"python3", *Language::Python.setup_install_args(prefix)
+      end
+    end
+    (libexec/"bin").install Dir["*"]
+    bin.install_symlink libexec/"bin/litani"
+
+    cd libexec/"bin/doc" do
+      system Formula["python@3.9"].opt_bin/"python3", "configure"
+      system "ninja"
+    end
+    man.install Dir[libexec/"bin/doc/out/man/*.roff"]
+    doc.install Dir[libexec/"bin/doc/out/html/index.html"]
```
```suggestion
    pythonpath = libexec/"vendor"/Language::Python.site_packages("python3")
    ENV.prepend_create_path "PYTHONPATH", pythonpath
    resources.each do |r|
      r.stage do
        system "python3", *Language::Python.setup_install_args(libexec/"vendor")
      end
    end
    libexec.install Dir["*"]
    (bin/"litani").write_env_script libexec/"litani", PYTHONPATH: pythonpath

    cd libexec/"doc" do
      system "python3", "configure"
      system "ninja", "--verbose"
    end
    
    man.install Dir[libexec/"doc/out/man/*.roff"]
    doc.install Dir[libexec/"doc/out/html/index.html"]
```
Also, do we really need to install the entire `doc` directory? It seems like we don't want everything in it, and only the built products.

I suspect something like this should work:
```suggestion
    pythonpath = libexec/"vendor"/Language::Python.site_packages("python3")
    ENV.prepend_create_path "PYTHONPATH", pythonpath
    resources.each do |r|
      r.stage do
        system "python3", *Language::Python.setup_install_args(libexec/"vendor")
      end
    end
    libexec.install Dir["*"] - ["doc"]
    (bin/"litani").write_env_script libexec/"litani", PYTHONPATH: pythonpath

    cd "doc" do
      system "python3", "configure"
      system "ninja", "--verbose"
    end
    
    man.install Pathname.glob("doc/out/man/*.roff")
    doc.install "doc/out/html/index.html"
```

carlocab reviewed (commented):
Thanks; this is looking much better.

But, as mentioned previously: we **should not** be installing Python modules into the formula's prefix.

Another issue is that your repo doesn't actually meet our notability requirements.

Finally, if the tool is named `litani`, why is the repo called `aws-build-accumulator`?

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,69 @@
+  end
+
+  def install
+    xy = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
+    ENV.prepend_create_path "PYTHONPATH", lib/"python#{xy}/site-packages"
+    resources.each do |r|
+      r.stage do
+        system Formula["python@3.9"].opt_bin/"python3", *Language::Python.setup_install_args(prefix)
+      end
+    end
+    (libexec/"bin").install Dir["*"]
+    bin.install_symlink libexec/"bin/litani"
+
+    cd libexec/"bin/doc" do
+      system Formula["python@3.9"].opt_bin/"python3", "configure"
+      system "ninja"
+    end
+    man.install Dir[libexec/"bin/doc/out/man/*.roff"]
+    doc.install Dir[libexec/"bin/doc/out/html/index.html"]
```
Actually, this might be simpler. Apologies for overwhelming you with suggestions. This third one is probably ideal.
```suggestion
    venv = virtualenv_create(libexec/"vendor", "python3")
    venv.pip_install resources

    libexec.install Dir["*"] - ["doc"]
    (bin/"litani").write_env_script libexec/"litani", PATH: "\"#{libexec}/vendor/bin:${PATH}\""

    cd "doc" do
      system libexec/"vendor/bin/python3", "configure"
      system "ninja", "--verbose"
    end

    man.install Pathname.glob("doc/out/man/*.roff")
    doc.install "doc/out/html/index.html"
```

ronakfof replied:
Thank you for all the suggestions and help. You have been really responsive and helpful.
I made the change you suggested and also added
```
ENV.prepend_create_path "PYTHONPATH", libexec/"vendor"/Language::Python.site_packages("python3")
````
as `system "ninja", "--verbose"` seems to be failing without it with a `jinja2 missing` error.

----

One question I have is if i update Python version from `@3.9` to `@3.10`, it seems to be failing with the `missing jinja2 error` as well. What should I change to make sure that doesn't happen.

----

Also seems like `gnuplot` has a dependency of python@3.9 which is conflicting if i install `python@3.10` along with it. How can i get past it?

carlocab replied:
Instead of prepending to `PYTHONPATH`:
```ruby
ENV.prepend_path "PATH", libexec/"vendor/bin"
```
That way you're using your venv's interpreter.

carlocab replied:
> Also seems like `gnuplot` has a dependency of python@3.9 which is conflicting if i install `python@3.10` along with it. How can i get past it?

You can either drop the `gnuplot` dependency, stay with `python@3.9`, or help migrate `gnuplot` to `python@3.10`. Migrating `gnuplot` may require migrating many other formulae.

ronakfof replied:
> Instead of prepending to `PYTHONPATH`:
> 
> ```ruby
> ENV.prepend_path "PATH", libexec/"vendor/bin"
> ```
> 
> That way you're using your venv's interpreter.

Will test it and update. Thanks

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,69 @@
+      system Formula["python@3.9"].opt_bin/"python3", "configure"
+      system "ninja"
+    end
+    man.install Dir[libexec/"bin/doc/out/man/*.roff"]
```
Also, I'm still *really* suspicious about this line. Are you sure it leads to usable manpages? It seems like it shouldn't, because whatever's installed doesn't even live in a `man1`, `man3`, etc. subdirectory.

ronakfof replied:
Yes you're right. I will fix this.

ronakfof replied:
Resolved

ronakfof commented:
> Thanks; this is looking much better.
> 
> But, as mentioned previously: we **should not** be installing Python modules into the formula's prefix.
> 
Fixed by your suggestion
> Another issue is that your repo doesn't actually meet our notability requirements.
> 
I am aware about the repo not meeting notability requirements. This PR is to get the formula done to make sure it ready when it passes the requirements. Till then we will be hosting it under our `AWS` tap.

> Finally, if the tool is named `litani`, why is the repo called `aws-build-accumulator`?

Our [Repo](https://github.com/awslabs/aws-build-accumulator/) is hosted under our official GitHub organization of [awslabs](https://github.com/awslabs/) and it has a naming convention for repo names which needs to be followed for any open-source projects.

ronakfof commented:
Hi, I updated our code to generate correct man pages like you suggested.
I added the following int the Formula - 
```
    man1.install libexec.glob("doc/out/man/*.1")
    man5.install libexec.glob("doc/out/man/*.5")
    man7.install libexec.glob("doc/out/man/*.7")
```
Which seems to be working fine.
One way to do this using a loop is by - 
```
    libexec.glob("doc/out/man/*") do |file|
      eval("man#{File.extname(file).delete(".")}.install file")
    end
```
Which works too but seems like `brew` doesn't allow the usage of `eval`.
Is there any alternative to this or should i just do this?
>     man1.install libexec.glob("doc/out/man/*.1")
>     man5.install libexec.glob("doc/out/man/*.5")
>     man7.install libexec.glob("doc/out/man/*.7")

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+
+  def install
+    ENV.prepend_path "PATH", libexec/"vendor/bin"
+    venv = virtualenv_create(libexec/"vendor", Formula["python@3.9"].opt_bin/"python3")
```
```suggestion
    venv = virtualenv_create(libexec/"vendor", "python3")
```
Do we really need to be explicit about `python@3.9` here? This makes changing the `python@3.x` dependency harder later on, since it's easy to forget to change these references.

ronakfof replied:
Resolved

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+    venv = virtualenv_create(libexec/"vendor", Formula["python@3.9"].opt_bin/"python3")
+    venv.pip_install resources
+
+    libexec.install Dir["*"]
```
```suggestion
    libexec.install Dir["*"] - ["doc", "test", "examples"]
```
It looks like we can skip installing these if we're just removing them later.

ronakfof replied:
Resolved

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+    libexec.install Dir["*"]
+    (bin/"litani").write_env_script libexec/"litani", PATH: "\"#{libexec}/vendor/bin:${PATH}\""
+
+    cd libexec/"doc" do
```
If we're not going to install `doc` into `libexec`:
```suggestion
    cd "doc" do
```

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+      system libexec/"vendor/bin/python3", "configure"
+      system "ninja", "--verbose"
+    end
+    # libexec.glob("doc/out/man/*") do |file|
+    #   eval("man#{File.extname(file).delete(".")}.install file")
+    # end
```
```suggestion
```
We don't need to leave this comment here. (`send` instead of `eval` will work here, but the explicit calls below are much clearer.)

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+    # libexec.glob("doc/out/man/*") do |file|
+    #   eval("man#{File.extname(file).delete(".")}.install file")
+    # end
+    man1.install libexec.glob("doc/out/man/*.1")
+    man5.install libexec.glob("doc/out/man/*.5")
+    man7.install libexec.glob("doc/out/man/*.7")
+    doc.install libexec/"doc/out/html/index.html"
+    rm_rf [libexec/"doc", libexec/"test", libexec/"examples"]
```
If these won't be in `libexec`:
```suggestion
    man1.install buildpath.glob("doc/out/man/*.1")
    man5.install buildpath.glob("doc/out/man/*.5")
    man7.install buildpath.glob("doc/out/man/*.7")
    doc.install "doc/out/html/index.html"
```

carlocab commented on Formula/litani.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  test do
+    system "litani", "init", "--project-name", "test-installation", "--stages", "codegen", "verify", "check"
```
```suggestion
    system bin/"litani", "init", "--project-name", "test-installation", "--stages", "codegen", "verify", "check"
```
Let's be sure we're testing the right `litani` executable instead of relying on `PATH`. Same below.

ronakfof replied:
Resolved

carlocab reviewed (commented):
Thanks, @ronakfof. Just a few more suggestions, but this is looking close to ready for when your repo meets our notability requirement.

ronakfof commented:
Thank You @carlocab for all the help and suggestions.

ronakfof commented:
Hi @carlocab,
I would like to Thank you first for all the help and suggestions with the Litani formula.
I was working on brew formulas for 2 other projects - 
- https://github.com/awslabs/aws-viewer-for-cbmc
- https://github.com/model-checking/kani

And I ask that [Litani](https://github.com/awslabs/aws-build-accumulator), [cbmc-viewer](https://github.com/awslabs/aws-viewer-for-cbmc), and [kani](https://github.com/model-checking/kani) be added to [homebrew-core](https://github.com/Homebrew/homebrew-core).  They currently fail to satisfy the notability threshold for core, and I ask for exceptions to the threshold for the following reasons:

- These 3 projects are very much notable and have an active usage.
- Have an active team working behind them which assures regular updates in future.

- Here's a link to the main Rust blog where [Kani](https://github.com/model-checking/kani) got called out as well - https://blog.rust-lang.org/inside-rust/2022/02/22/compiler-team-ambitions-2022.html#librarification-aspirations-%EF%B8%8F

- [cbmc-viewer](https://github.com/awslabs/aws-viewer-for-cbmc) is used to view reports and [Litani](https://github.com/awslabs/aws-build-accumulator) is used to run proofs across all our current projects using [CBMC](https://github.com/diffblue/cbmc) and already published on Brew [[HERE](https://formulae.brew.sh/formula/cbmc)].

I am requesting for these projects to be considered to be merged without the notability requirements on basis of an active team working behind them and their active usage and customer base.

Right now we have this Pull Request for Litani formula and soon we'll be creating pull requests for [Kani](https://github.com/model-checking/kani) and [cbmc-viewer](https://github.com/awslabs/aws-viewer-for-cbmc).

If there is any additional information required by you in order to make this decision, please let me know.

carlocab commented:
Do you have any install/usage statistics for these proposed formulae?

Note that these formulae do not need to be hosted in homebrew-core for them to be installable using Homebrew. You might wish to consider hosting them in your own tap in the meantime.

Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

ronakfof commented:
> Do you have any install/usage statistics for these proposed formulae?

Hi @carlocab,
Here is a list of open-source projects which uses `litani` and `cbmc-viewer` actively - 
- [Fleet-Provisioning-for-AWS-IoT-embedded-sdk](https://github.com/aws/Fleet-Provisioning-for-AWS-IoT-embedded-sdk)
- [FreeRTOS-Plus-TCP](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP)
- [FreeRTOS](https://github.com/FreeRTOS/FreeRTOS)
- [Lab-Project-FreeRTOS-Cellular-Library](https://github.com/FreeRTOS/Lab-Project-FreeRTOS-Cellular-Library)
- [SigV4-for-AWS-IoT-embedded-sdk](https://github.com/aws/SigV4-for-AWS-IoT-embedded-sdk)
- [amazon-freertos](https://github.com/aws/amazon-freertos)
- [aws-c-common](https://github.com/awslabs/aws-c-common)
- [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)
- [coreHTTP](https://github.com/FreeRTOS/coreHTTP)
- [coreJSON](https://github.com/FreeRTOS/coreJSON)
- [coreMQTT-Agent](https://github.com/FreeRTOS/coreMQTT-Agent)
- [coreMQTT](https://github.com/FreeRTOS/coreMQTT)
- [corePKCS11](https://github.com/FreeRTOS/corePKCS11)
- [coreSNTP](https://github.com/FreeRTOS/coreSNTP)
- [device-defender-for-aws-iot-embedded-sdk](https://github.com/aws/device-defender-for-aws-iot-embedded-sdk)
- [device-shadow-for-aws-iot-embedded-sdk](https://github.com/aws/device-shadow-for-aws-iot-embedded-sdk)
- [jobs-for-aws-iot-embedded-sdk](https://github.com/aws/jobs-for-aws-iot-embedded-sdk)
- [ota-for-aws-iot-embedded-sdk](https://github.com/aws/ota-for-aws-iot-embedded-sdk)
- [s2n-tls](https://github.com/aws/s2n-tls)

These projects have been using `Litani` as a submodule and we already have a PR out which will remove it from the repository and will have to be used as an installation binary from brew package.
That means, all the SDEs working on all these projects will be using `litani's` and `cbmc-viewer's` brew package to set-up the environment.

Along with this, we have private projects with similar usage and we will be onboarding new projects as well.

> Note that these formulae do not need to be hosted in homebrew-core for them to be installable using Homebrew. You might wish to consider hosting them in your own tap in the meantime.
> 

I am aware about using or creating external taps and we have been using [aws/tap](https://github.com/aws/homebrew-tap) for a while now. But it has some limitations such as testing, bottle creation etc and it's not ideal to create more taps under same organization.

carlocab reviewed (approved):
I'm ok with merging this one, but I'd like to wait and see what the install analytics (`brew info --analytics litani`) are like before adding the other formulae you're proposing.

ronakfof commented:
Hi @carlocab,
Thank you for all the assistance, suggestions and prompt review.
This would be really helpful for our team and customers.
